src/klein/_headers.py

```diff
@@ -0,0 +1,315 @@
+        Raw header data as a tuple in the from: C{((name, value), ...)}.
+        C{name} and C{value} are bytes.
+        Headers are provided in the order that they were received.
+        Headers with multiple values are provided as separate name and value
+        pairs.
+        """
+    )  # type: RawHeaders
+
+
+    def getValues(name):
+        # type: (AnyStr) -> Iterable[AnyStr]
+        """
+        Get the values associated with the given header name.
+
+        If the given name is L{bytes}, the value will be returned as the raw
+        header L{bytes}.
+
+        If the given name is L{Text}, the name will be encoded as ISO-8859-1
+        and the value will be returned as text, by decoding the raw header
+        value bytes with ISO-8859-1.
+
+        @param name: The name of the header to look for.
+
+        @return: The values of the header with the given name.
+        """
+
+
+
+class IMutableHTTPHeaders(IHTTPHeaders):
+    """
+    Mutable HTTP entity headers.
+    """
+
+    def remove(name):
+        # type: (AnyStr) -> None
+        """
+        Remove all header name/value pairs for the given header name.
+
+        If the given name is L{Text}, it will be encoded as ISO-8859-1 before
+        comparing to the (L{bytes}) header names.
+
+        @param name: The name of the header to remove.
+        """
+
+
+    def addValue(name, value):
+        # type: (AnyStr, AnyStr) -> None
+        """
+        Add the given header name/value pair.
+
+        If the given name is L{bytes}, the value must also be L{bytes}.
+
+        If the given name is L{Text}, it will be encoded as ISO-8859-1, and the
+        value, which must also be L{Text}, will be encoded as ISO-8859-1.
+        """
+
+
+
+# Encoding/decoding header data
+
+HEADER_NAME_ENCODING  = "iso-8859-1"
+HEADER_VALUE_ENCODING = "iso-8859-1"
+
+
+def headerNameAsBytes(name):
+    # type: (String) -> bytes
+    """
+    Convert a header name to bytes if necessary.
+    """
+    if isinstance(name, bytes):
+        return name
+    else:
+        return name.encode(HEADER_NAME_ENCODING)
+
+
+def headerNameAsText(name):
+    # type: (String) -> Text
+    """
+    Convert a header name to text if necessary.
+    """
+    if isinstance(name, Text):
+        return name
+    else:
+        return name.decode(HEADER_NAME_ENCODING)
+
+
+def headerValueAsBytes(value):
+    # type: (String) -> bytes
+    """
+    Convert a header value to bytes if necessary.
+    """
+    if isinstance(value, bytes):
+        return value
+    else:
+        return value.encode(HEADER_VALUE_ENCODING)
+
+
+def headerValueAsText(value):
+    # type: (String) -> Text
+    """
+    Convert a header value to text if necessary.
+    """
+    if isinstance(value, Text):
+        return value
+    else:
+        return value.decode(HEADER_VALUE_ENCODING)
+
+
+def normalizeHeaderName(name):
+    # type: (AnyStr) -> AnyStr
+    """
+    Normalize a header name.
+    """
+    return name.lower()
```

Comment:
I feel like this should also commit to returning bytes.

Reply:
True if we change the API to remove the magical typing noted above